ENH: add block() function to create block arrays

sotte opened the pull request:
`block` is similar to matlab's "square bracket stacking" functionality: `[A B; C D]`.

So you can write:

``` python
>>> A = np.array([[1, 2, 3]])
>>> B = np.array([[2, 3, 4, 5, 6, 7]])
>>> c = np.array([8, 3, 4, 5, 6, 7])
>>> d1 = np.array([9])
>>> d2 = np.array([8, 7, 6, 5, 4])
>>> block([[A, A],
           [B],
           [c],
           [d1, d2]])
array([[1, 2, 3, 1, 2, 3],
       [2, 3, 4, 5, 6, 7],
       [8, 3, 4, 5, 6, 7],
       [9, 8, 7, 6, 5, 4]])
```

The one thing I'm not super happy with is that 1-D arrays are treated as row vectors and I personally think of them as column vectors.

Are the doctests sufficient or should I create proper unittests?

sotte committed:
ENH: add block() function.

`block` is similar to Matlab's square bracket stacking functionality for block
matices. `block` is an addition to the current stacking functions vstack,
hstack, stack.

abalkin commented:
> The one thing I'm not super happy with is that 1-D arrays are treated as row vectors and I personally think of them as column vectors.

Why?

They sure look like rows when printed:

In [8]: np.arange(10)
Out[8]: array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

sotte commented:
I updated my PR. Because of the discussion here #5605 I renamed the function to `block` because you basically create a block matrix. I moved the tests into a proper TestCase.

Some remarks:
- I also played with the implementation of `bmat` as skeleton for `block`. But the implementation is too complex for my liking. And `block` in the current form does everything I want.
- On the mailinglist we discussed more advanced versions of `block` but I think the current version actually versatile enough.

Feedback welcome!

sotte commented:
@abalkin it's just my personal pet I guess :) It does not really matter for the pull request.

sotte commented:
Do you think this is ready to be merged?

sotte commented:
One of my new year's resolution is to get my pull requests merged. So, what do you think?

rgommers commented:
> One of my new year's resolution is to get my pull requests merged. So, what do you think?

Sorry that this didn't get reviewed yet Stefan. It seems useful to multiple people and isn't hard to review, so let's make sure we have a serious look before the next release (I gave it the 1.12.0 milestone).

rgommers commented on numpy/core/shape_base.py:
```diff
@@ -348,3 +353,83 @@ def stack(arrays, axis=0):
+    """
+    Create block arrays similar to Matlab's "square bracket stacking":
+
+            [A A; B B]
```
Can you hide this somewhere in the Notes section? Putting Matlab syntax here so prominently feels wrong, plus for a minute I read this as the Python syntax to use.

rgommers commented on numpy/core/shape_base.py:
```diff
@@ -348,3 +353,83 @@ def stack(arrays, axis=0):
+
+    Parameters
+    ----------
+    tup_tup : sequence of sequence of ndarrays
```
sequence of sequence is quite ugly. Looking at the examples, why would it be needed to use a list of a single array, instead of that array itself? To clarify, in this example

    >>> block([[One, Two],
    ...        [Three],
    ...        [four],

why not write that as


    >>> block([[One, Two],
    ...        Three,
    ...        four,

?

rgommers replied:
Also, ``tup_tup`` is not a very clear name. Maybe just call it ``arrays``, like ``stack`` does?

sotte replied:
I didn't want to use `*args` because a) I don't like `*args`, b) looking at the other pull requests here `*args` seems to be discouraged, c) you can't add any other arguments (excecpt `**kwargs` which is also not great) and on the mailinglist there was some discussion on additotional parameters, and d) block behaves exactly like `array` and everybody is familiar with that. But I'm open for changes.

I agree, `tup_tup` can have a better name.

rgommers commented on numpy/core/shape_base.py:
```diff
@@ -348,3 +353,83 @@ def stack(arrays, axis=0):
+
+    Examples
+    --------
+    Stacking in a row:
```
A blank line is needed here, or it won't render correctly.

sotte replied:
All other function in `shape_base.py` are exactly like this. I'll build the docs and check..

rgommers commented on numpy/core/shape_base.py:
```diff
@@ -348,3 +353,83 @@ def stack(arrays, axis=0):
+    >>> block([a, b])
+    array([[1, 1, 2, 2]])
+    """
+    if isinstance(tup_tup[0], list) or isinstance(tup_tup[0], tuple):
```
shorter: ``isinstance(tup_tup[0], (tuple, list))``

sotte replied:
Every time I see this I think it's cool but then I forget. Will change it.

rgommers commented on numpy/core/tests/test_shape_base.py:
```diff
@@ -315,5 +318,71 @@ def test_stack():
+        assert_almost_equal(block([[a, b], [a, b]]),
+                            np.array([[1, 2, 3, 2, 3, 4],
+                                      [1, 2, 3, 2, 3, 4]]))
+
```
I'm missing a few tests for bad-weather behavior: empty arrays, and arrays with mismatched shapes.

sotte replied:
Good idea. I found a bug. I would have expected this to throw an exception.

```
     >>> a = np.array([0, 0])
    >>> b = np.eye(2)
    >>> np.block(((a, a), b))
    array([[0, 0. 0, 0],
           [1, 0, 0, 1]])
```

`b` is `hstack`ed to `[1, 0, 0, 1]`.

sotte committed:
Move comment about matlab notation comment to Note section.

sotte committed:
rename tup_tup to arrays

sotte committed:
Simplify if

sotte committed:
Better examples in docstring

sotte commented:
I can't see the forest for the trees so I talked to my rubber duck.

What do we expect? Do we want `block([A, B])` to be equivalent to `vstack([A, B])`, i.e., should A and B be stacked in a row? Of course, right? But what about `block([A, [B, C]])` then? On first glance this looks like the first example. Do we have to analyse the complete structure of the nested arrays? Should we throw an Exception there? Maybe we should enforce explicit brackets?
`block([[A, B]])`  and `block([[A], [B, C]]` are clear,  but they are verbose. 

If we use `*args` it's less verbose `block(A, [B, C])` and I think it's also clear that `block([A, B])` is equivalent to `vstack([A, B])`. (Also, when using `*args` the implementation is clean because `*args` is always a tuple where each element represents a row.)

Now, I'm actually inclined to use `*args` despite my general dislike of it.

sotte committed:
*args version of block

sotte committed:
Unittest for unneeded []

charris commented:
@sotte Take a look at the commit message guidelines: `doc/source/dev/gitwash/development_workflow.rst`. You also need to squash your commits and redo the message(s), you can use `git rebase -i HEAD~7` for that followed by a force push `git push -f origin HEAD`.

charris commented:
Updated in #7768, so closing this. Thanks @sotte .

sotte commented:
@charris I guess I'm too late. Thanks a lot! Was a lot of fun...as always!
